enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreStateApplier.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.coreedge.raft.state;
+
+import java.util.concurrent.ExecutorService;
+import java.util.function.Function;
+
+import org.neo4j.helpers.NamedThreadFactory;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+
+import static java.util.concurrent.Executors.newSingleThreadExecutor;
+import static java.util.concurrent.TimeUnit.HOURS;
+
+/**
+ * Supplies an executor service suitable for the core state.
+ * Running tasks can be cancelled using a soft abort mechanism.
+ */
+public class CoreStateApplier
+{
+    public class Status
+    {
+        private boolean cancelled;
```

Comment:
There is no race because it is just a soft-abort mechanism. The intention is to get the applier to eventually stop, it is fine if it does it within 1 second or 1 minute, just eventually.